Refactor section restriction callback by using a hook

PhMemmel opened the pull request:
This should rework the already merged #99 and replace #117

PhMemmel commented:
@Syxton I would be happy about your input on this so far, especially about the design of the hook class. There of course are plenty of possibilities how to design them, for example only offer one method to remove restricted sections like I am doing currently, or offer a method to filter sections etc. Do you have any thoughts on this?

Syxton commented on classes/hook/filter_sections.php:
```diff
@@ -0,0 +1,193 @@
+    private readonly array $originalsectionnums;
+
+    /** @var bool Determines if the user will be able to keep the original section of a course module when performing some operations. */
+    private bool $keeporiginalsectionallowed;
```
As a typed property, does this need to be initialized somewhere?  I would think the is_keeporiginalsectionallowed() method would fail on actions in SAMECOURSE.  Should it be set to false by default?

PhMemmel replied:
Yeah, this was because of the combined designed of both hooks in one. I changed that design but also included your feedback.

Syxton commented on classes/hook/filter_sections.php:
```diff
@@ -0,0 +1,193 @@
+    private bool $keeporiginalsectionallowed;
+
+    /** @var bool Determines if the user will be able to create a new section when performing some operations. */
+    private bool $createnewsectionallowed;
```
Same question as above with an uninitialized typed property.

Syxton reviewed (commented):
I really like how this works.  With how rare its use case is, I think this will work nicely to fulfill the need to filter the section lists.

Syxton commented:
@PhMemmel I think I want to have this in place for the 4.4 release.  So I'll wait to create the release.  But if we get close to the 10th, I'll make the release and we can pull this in later. Gotta have that badge :)

PhMemmel commented:
We got this in time, no worries. Will probably manage to take care of your feedback and finalize this on Monday 👍

PhMemmel commented:
Hi @Syxton,

I took another look and decided to rework the hook by splitting it up into two hooks. Reason: Hooks are some kind of external api without being well documented ;-), so they should be easy to understand and use. The current approach was a bit complicated, so I split it up. To avoid code duplication one is tempted to use inheritance to avoid this, however this is discouraged in the moodle hooks docs, so I used a trait. I'm not 100% happy with that, but it still lets me sleep peacefully... ;-)

Let me know, what you think. For testing I also updated the branch with my changes to format_topics (it's moodle 4.5 current develop now).

Syxton commented:
@PhMemmel Sorry it took so long to approve this.  I ended up getting deep in the weeds about how Moodle implemented hooks.  It is my understanding that Moodle hooks are more like events, but they called them hooks because events in Moodle are already a thing.  Anyway.  It makes good sense to me to split those and the trait and hidden input really are decent clean code solution.  If you are good with where this request is, I'll merge and get a 4.4+ release out today.

PhMemmel commented:
Hi @Syxton, first of all: that's definitely not a long time :)

Besides that thank you very much for the feedback. If you have no more, this is ready to integrate from my point of view.
